EFS-136 Retry failed Commitlog sends

gcatlin opened the pull request:
[EFS-136: Build PCH retry logic](https://spreedly.atlassian.net/browse/EFS-136)

In AWS, we're running Commitlog separate from Riak such that postcommit-hook now must make a network call to reach Commitlog. This can fail due to a network partition, Commitlog could be down, Commitlog could be at capacity, etc. Additionally, we see a small number of failures occurring during blue-green deployments of Commitlog. So we need PCH to (A) handle temporary unavailability of Commitlog and (B) do no worse than we do now for longer periods of unavailability.

This PR addresses A, but B will be addressed by a separate PR pending feedback on this one.

Example logging produced by this PR from 7 concurrent requests, with Commitlog stopped initially, then started during retries:
```
2019-06-21 00:35:21.507 [warning] <0.20226.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:     1 ms. Retries: 0. test_bucket|a
2019-06-21 00:35:21.528 [warning] <0.20234.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:     1 ms. Retries: 0. test_bucket|b
2019-06-21 00:35:21.554 [warning] <0.20246.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:     1 ms. Retries: 0. test_bucket|c
2019-06-21 00:35:21.589 [warning] <0.20262.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:     3 ms. Retries: 0. test_bucket|d
2019-06-21 00:35:21.606 [warning] <0.20226.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   100 ms. Retries: 1. test_bucket|a
2019-06-21 00:35:21.609 [warning] <0.20271.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:     2 ms. Retries: 0. test_bucket|e
2019-06-21 00:35:21.626 [warning] <0.20278.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:     1 ms. Retries: 0. test_bucket|f
2019-06-21 00:35:21.635 [warning] <0.20234.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   109 ms. Retries: 1. test_bucket|b
2019-06-21 00:35:21.644 [warning] <0.20287.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:     1 ms. Retries: 0. test_bucket|g
2019-06-21 00:35:21.652 [warning] <0.20246.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   100 ms. Retries: 1. test_bucket|c
2019-06-21 00:35:21.691 [warning] <0.20262.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   105 ms. Retries: 1. test_bucket|d
2019-06-21 00:35:21.709 [warning] <0.20271.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   102 ms. Retries: 1. test_bucket|e
2019-06-21 00:35:21.724 [warning] <0.20278.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   100 ms. Retries: 1. test_bucket|f
2019-06-21 00:35:21.743 [warning] <0.20287.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   101 ms. Retries: 1. test_bucket|g
2019-06-21 00:35:21.808 [warning] <0.20226.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   302 ms. Retries: 2. test_bucket|a
2019-06-21 00:35:21.820 [warning] <0.20234.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   294 ms. Retries: 2. test_bucket|b
2019-06-21 00:35:21.822 [warning] <0.20246.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   270 ms. Retries: 2. test_bucket|c
2019-06-21 00:35:21.886 [warning] <0.20262.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   301 ms. Retries: 2. test_bucket|d
2019-06-21 00:35:21.899 [warning] <0.20271.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   293 ms. Retries: 2. test_bucket|e
2019-06-21 00:35:21.908 [warning] <0.20278.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   283 ms. Retries: 2. test_bucket|f
2019-06-21 00:35:21.954 [warning] <0.20287.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   311 ms. Retries: 2. test_bucket|g
2019-06-21 00:35:22.124 [warning] <0.20246.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   572 ms. Retries: 3. test_bucket|c
2019-06-21 00:35:22.146 [warning] <0.20234.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   619 ms. Retries: 3. test_bucket|b
2019-06-21 00:35:22.236 [warning] <0.20278.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   611 ms. Retries: 3. test_bucket|f
2019-06-21 00:35:22.246 [warning] <0.20271.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   639 ms. Retries: 3. test_bucket|e
2019-06-21 00:35:22.255 [warning] <0.20226.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   749 ms. Retries: 3. test_bucket|a
2019-06-21 00:35:22.296 [warning] <0.20262.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   710 ms. Retries: 3. test_bucket|d
2019-06-21 00:35:22.396 [warning] <0.20287.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:   754 ms. Retries: 3. test_bucket|g
2019-06-21 00:35:22.735 [warning] <0.20246.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  1183 ms. Retries: 4. test_bucket|c
2019-06-21 00:35:22.805 [warning] <0.20234.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  1278 ms. Retries: 4. test_bucket|b
2019-06-21 00:35:22.927 [warning] <0.20278.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  1302 ms. Retries: 4. test_bucket|f
2019-06-21 00:35:22.959 [warning] <0.20271.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  1352 ms. Retries: 4. test_bucket|e
2019-06-21 00:35:23.096 [warning] <0.20262.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  1510 ms. Retries: 4. test_bucket|d
2019-06-21 00:35:23.111 [warning] <0.20226.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  1605 ms. Retries: 4. test_bucket|a
2019-06-21 00:35:23.194 [warning] <0.20287.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  1551 ms. Retries: 4. test_bucket|g
2019-06-21 00:35:23.829 [warning] <0.20246.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  2276 ms. Retries: 5. test_bucket|c
2019-06-21 00:35:24.061 [warning] <0.20234.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  2534 ms. Retries: 5. test_bucket|b
2019-06-21 00:35:24.281 [warning] <0.20278.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  2656 ms. Retries: 5. test_bucket|f
2019-06-21 00:35:24.292 [warning] <0.20271.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  2685 ms. Retries: 5. test_bucket|e
2019-06-21 00:35:24.680 [warning] <0.20262.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  3094 ms. Retries: 5. test_bucket|d
2019-06-21 00:35:24.885 [warning] <0.20226.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  3379 ms. Retries: 5. test_bucket|a
2019-06-21 00:35:24.895 [warning] <0.20287.0> [commitlog] Unable to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  3252 ms. Retries: 5. test_bucket|g
2019-06-21 00:35:26.025 [warning] <0.20246.0> [commitlog] FAILED to sync data to commitlog: {nodedown,'commitlog@127.0.0.1'}. Time:  4473 ms. Retries: 6. test_bucket|c
2019-06-21 00:35:26.680 [info] <0.20234.0>    [commitlog] sync_to_commitlog success. Time:  5153 ms. Retries: 6. test_bucket|b
2019-06-21 00:35:26.726 [info] <0.20271.0>    [commitlog] sync_to_commitlog success. Time:  5119 ms. Retries: 6. test_bucket|e
2019-06-21 00:35:27.348 [info] <0.20278.0>    [commitlog] sync_to_commitlog success. Time:  5723 ms. Retries: 6. test_bucket|f
2019-06-21 00:35:28.008 [info] <0.20262.0>    [commitlog] sync_to_commitlog success. Time:  6422 ms. Retries: 6. test_bucket|d
2019-06-21 00:35:28.092 [info] <0.20226.0>    [commitlog] sync_to_commitlog success. Time:  6586 ms. Retries: 6. test_bucket|a
2019-06-21 00:35:28.224 [info] <0.20287.0>    [commitlog] sync_to_commitlog success. Time:  6581 ms. Retries: 6. test_bucket|g
```

rwdaigle commented:
May or may not be relevant but we do have [gentry](https://github.com/spreedly/gentry) for generic retry logic?

gcatlin commented:
Interesting. I'll take a look. Thanks!

gcatlin commented on src/postcommit_hook.erl:
```diff
@@ -46,6 +67,48 @@ sync_to_commitlog(Action, Bucket, Key, Value) ->
+            retries_exceeded;
+        E ->
+            E
+    end.
```
Does this seem reasonable? Is there a more Erlangy alternative to `sleep`?

dtykocki replied:
It's pretty uncommon to to use `sleep` in both Elixir and Erlang (well other than async tests). This'll suspend the process itself which may be fine in this context, but it'd suggest using giving http://erlang.org/doc/man/erlang.html#send_after-3 a try here. This'll send a message to process a process (in this case `self`) after N milliseconds.

dtykocki commented:
> May or may not be relevant but we do have [gentry](https://github.com/spreedly/gentry) for generic retry logic?

I would shy away from using an Elixir library in an Erlang application. Here's my thinking:
1. Calling Elixir from Erlang is kind of awkward and not very common. The module and function being called must be wrapped in single quotes: `'Elixir.IO':'puts'("Hello, world!").`
2. Since we're in the PCH, we're tied to a really old version of Erlang (I think R16 something). It's entirely possible the PCH may not even compile once an Elixir lib is added.
3. Given the importance and simplicity of the PCH, I'd strive to keep the number of dependency to a minimum - especially for something simple like retries. If we're to pull in a dependency, I'd default to an Erlang library first. Maybe something like https://github.com/ferd/backoff.

gcatlin commented:
Yeah, I was looking at `ferd/backoff` but it didn't seem to add enough value to make it worth having an additional dependency.

dtykocki commented on src/postcommit_hook.erl:
```diff
@@ -1,39 +1,60 @@
+    % send_with_retries(Object, timestamp(), 0, ?INITIAL_RETRY_DELAY_MS)
+    Sender = spawn(fun() -> sender() end),
+    Sender ! {self(), Object},
+    ok.
```
Hmm this seems unnecessary and filled with non-obvious edge cases at first glance. By using `spawn` we'll have to account for situations in which a process crashes, is slow to respond, or never returns a reply back. Additionally, we'll no longer put back pressure on the caller (Riak in this case?) which could lead to spawning an unbounded number of processes, eventually leading to a VM crash (worst case scenario).

Some of these cases can be handled by using `spawn_monitor` in combination with a well crafted `receive` block but I do question introducing this level of complexity. What are we really gaining by using processes here?

gcatlin replied:
Exponential backoffs are a significant change in behavior, so I was thinking using a separate process would somewhat preserve the existing behavior. Specifically, I'm concerned with what happens when Commitlog is unavailable for more than a few seconds and many PCHs pile up as they all wait for the retry process to complete. But I agree with your points about added complexity and back pressure.

Maybe removing the `spawn` and implementing some version of the circuit-breaker-ish functionality mentioned below is the better way to address the "Commitlog is down" scenario?

gcatlin replied:
@objectuser

>At that point we could leverage OTP ...

Since PCH is a dynamic library, not an application, I'm not sure how to leverage OTP design patterns. For example, how do we `start` a supervisor?

> is the PCH getting a new process every time?
> Is the call to PCH sync or async from the perspective of Riak?

I found a [good summary of how it works](https://github.com/basho/riak_kv/blob/develop-2.9/docs/Node-Diversity.md#implementation). In short, it looks like each Riak `put` is run *synchronously* in a [sidejob](https://github.com/basho/sidejob/) "resource" (inside a gen_server). By default, Riak supports [10,000](https://github.com/basho/riak_kv/blob/master/src/riak_kv_app.erl#L43) sidejob "resources".

The initial flow looks like this:
1. [riak_client:normal_put](https://github.com/basho/riak_kv/blob/develop-2.9/src/riak_client.erl#L201)
1. [riak_kv_put_fsm:start](https://github.com/basho/riak_kv/blob/develop-2.9/src/riak_kv_put_fsm.erl#L156)
1. [sidejob_supervisor:start_child](https://github.com/basho/sidejob/blob/master/src/sidejob_supervisor.erl#L52)
1. [sidejob:call](https://github.com/basho/sidejob/blob/master/src/sidejob.erl#L67)
1. Then, as far as I can tell, each state of the fsm is executed from there.

So, it appears that Riak will refuse work if all the sidejob resources are in use, which means the exponential backoff logic does have the potential to negatively impact Riak's availability.

> Is it receiving any other kinds of messages that we might clash with if we have a `receive` block here?

I'm not sure. I vaguely recall getting some unexpected-by-me messages in an earlier incarnation of this code. I'll explore it more in a separate circuit-breaker branch.

objectuser commented:
So, to summarize the drivers?
- The PCH -> Commitlog integration used to be pretty reliable due to them running on the same machine
- Now they'll be running on different machines in a cloud, so we anticipate more issues with that connectivity
- Additionally we see errors during deployments now, which are now much more automated/lower touch than they were before
- In this context the previous manual log-based recovery process will be too onerous with the expected frequency of failures.

With this change we'll still leverage the log-based recovery if we exhaust the retries, but we want to use it only under more extreme circumstances, not just when we do a deployment, experience brief network interruptions, etc.

objectuser commented on src/postcommit_hook.erl:
```diff
@@ -46,6 +67,48 @@ sync_to_commitlog(Action, Bucket, Key, Value) ->
+    end.
+
+%% TODO some kind of circuit breaker thing in case Commitlog is down for an
+%%      extended period of time.
```
This is a really good point. What are the implications on a Riak node if a put operation takes increasingly longer. Having some sort of circuit breaker functionality might be a good approach.

I'll say that I really don't have any idea if that's problematic or not, however. If the Riak commit happens async from firing the hook (no matter if "something" makes a blocking call to the PCH or not) then it might not matter at all.

Managing a circuit breaker across lots of Riak processes ... maybe that wouldn't be too hard in theory, at least within a single Riak node. I'm not sure how we initialize the process though.

Significant complexity creeping in at this point.

gcatlin commented:
Please review this PR one more time. You might want to hide comments on the PR to make it easier to read (`···` > ✓ Show Comments).

Changes:
 - removed the subprocess
 - replaced `[commitlog]` with `[postcommit-hook]` in the log messages

I'll create a separate PR soon for the circuit-breaker proposal.

gcatlin commented:
Follow PR: https://github.com/spreedly/riak-postcommit-hook/pull/19

objectuser commented on src/postcommit_hook.erl:
```diff
@@ -1,39 +1,60 @@
-            throw(ExtractException)
+            error_logger:warning_msg("[commitlog] Unable to extract data from Riak object: ~s. Object: ~s.",
+                                     [ExtractException, Object]),
+            throw({e_extract, ExtractException})
```
@gcatlin Do we have anything scraping the logs for `[commitlog]` that might now need to look for `[postcommit-hook]`?

objectuser commented on src/postcommit_hook.erl:
```diff
@@ -46,6 +67,52 @@ sync_to_commitlog(Action, Bucket, Key, Value) ->
+send_with_retries(Object, StartTime, RetryCount, RetryDelay) ->
+    try attempt_send_to_kafka_riak_commitlog(Object, StartTime, RetryCount)
+    catch
+        {e_sync, _} when RetryCount < ?MAX_RETRIES ->
```
Well it just popped into my head. I don't know if it's worth it really, or "simpler" but I somehow gravitate to overloaded function heads to remove conditionals. For example, if it started counting from `?MAX_RETRIES` down to zero and instead of `throw`s in `attempt_send_to_kafka_riak_commitlog`, it just returned the tuple, it could be something like this:

```erlang
%% First try
try_send(Object, StartTime, ?MAX_RETRIES, _) ->
    seed(),
    case attempt_send_to_kafka_riak_commitlog(Object, StartTime, RetryCount) of
        ok -> ok;
        ErrorStatus -> try_send(Object, StartTime, RetryCount - 1, ErrorStatus)
    end;
    
%% No more tries
try_send(_, _, 0, ErrorStatus) ->
    ErrorStatus;

%% Try after failure
try_send(Object, StartTime, RetryCount, _) ->
    RetryDelay = jitter((?MAX_RETRIES - RetryCount) * ?RETRY_DELAY_MULTIPLIER, ?RETRY_DELAY_JITTER),
    timer:sleep(RetryDelay),
    case attempt_send_to_kafka_riak_commitlog(Object, StartTime, RetryCount) of
        ok -> ok;
        ErrorStatus -> try_send(Object, StartTime, RetryCount - 1, RetryDelay, ErrorStatus)
    end.
    
%% Seeding is necessary to avoid a race condition where multiple postcommit-
%% hook processes automatically seed the RNG in the same manner and all use the
%% same seed, which defeats the purpose of adding jitter.
seed() ->
    {_, Seconds, MicroSecs} = now(),
    random:seed(erlang:phash2(self()), Seconds, MicroSecs).
```

That removes some embedded conditionals, including the overloaded `seed_if`. Seems like we could do even more extraction with the logging in `attempt_send_to_kafka_riak_commitlog` that checks `RetryCount < ?MAX_RETRIES`.

Like I said, just popped into my head when I was reading it. I suggest we move forward with what you have. 😊

gcatlin commented:
Closing in favor of #22
